### PR TITLE
added triple loss func to paddle frontend and its test

### LINKED ISSUE #20878: triplet_margin_loss

### PR DESCRIPTION
**Pull request by bundleabundance**
Close #20878

### CODE REVIEW

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 1]
```diff
@@ -210,3 +210,22 @@ def kl_div(
...
+    distance_negative = ivy.mean(ivy.square(anchor - negative), axis=-1, keepdims=True)
+    loss = ivy.relu(distance_positive - distance_negative + margin)
+
+    reduction = _get_reduction_func(reduction)
+    loss = reduction(loss)
+
+    if loss.shape == ():
+        loss = loss.expand_dims()
+
+    return paddle.to_tensor(ivy.atleast_1d(loss))
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:231]: we don't call the framework in the frontend function, only we use ivy functions and we return our custom frontend arrays.

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey @bundleabundance 

Thanks for contributing to ivy😊
Below are a few changes requested, kindly look into them and feel free to reachout if you face any problem🙂

Thanks

File: ivy/functional/frontends/paddle/nn/functional/loss.py (hunk 1 above)
Comment (bundleabundance) [reply ivy/functional/frontends/paddle/nn/functional/loss.py]: I am not sure I understand what should I do.

Comment (hirwa-nshuti) [reply ivy/functional/frontends/paddle/nn/functional/loss.py]: What I was saying is that we have frontend function wrappers that returns frontend Tensors we don't use any framework dependency on our frontends, we only use a composition of ivy functions

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey

We don't need paddle dependency in a frontend function, you can just use ivy functions

File: ivy/functional/frontends/paddle/nn/functional/loss.py (hunk 1 above)
Comment (bundleabundance) [reply ivy/functional/frontends/paddle/nn/functional/loss.py]: This has to be the correct way then I uploaded it, could you look at the tests and see if there is anything else to do?

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
I checked from paddle docs this function named this doesn't exist it is our mistake in writing down issue for contribution would you change this to implement `triplet_margin_loss` which exists in the paddle framework

**Comment by bundleabundance** [comment]
> 

Ok so what should I do now? This is my third attempt on a pull request and I can't seem to achieve a single one. I am really eager to attend to the unify internship but I don't know what to do.

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey @bundleabundance 

Some quick changes are needed then we will progress to see if the unit tests are running successfully🙂
Kindly check the comments below

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 2]
```diff
@@ -230,3 +230,56 @@ def margin_ranking_loss(input, other, label, margin=0.0, reduction="mean", name=
     out = ivy.atleast_1d(out)
 
     return out
+
+
+@to_ivy_arrays_and_back
+@with_supported_dtypes({"2.5.1 and below": ("float32", "float64")}, "paddle")
+def triplet_margin_loss(
```
[the remark this answers is not in the document]
Comment (bundleabundance) [reply ivy/functional/frontends/paddle/nn/functional/loss.py:237]: I looked at the documentation and it uses float32 and float64 as well, is there anything wrong? Argument names are same as well.

[the remark this answers is not in the document]
Comment (hirwa-nshuti) [reply ivy/functional/frontends/paddle/nn/functional/loss.py:237]: The first argument of the function is supposed to be `input`

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Check the updated comments also the tests are failing to the unused and undeclared fixtures in your unit test, would you try to run them locally for you to be able to debug the failures

**Comment by bundleabundance** [comment]
> Check the updated comments also the tests are failing to the unused and undeclared fixtures in your unit test, would you try to run them locally for you to be able to debug the failures

The individual tests are working now. The failing tests here are not because of my contribution/

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 3]
```diff
@@ -230,3 +230,56 @@ def margin_ranking_loss(input, other, label, margin=0.0, reduction="mean", name=
...
+    reduction = _get_reduction_func(reduction)
+
+    a_dim = input.ndim
+    p_dim = positive.ndim
+    n_dim = negative.ndim
+
+    ivy.assertions.check_true(
+        a_dim == p_dim and p_dim == n_dim,
+        lambda: (
+            "The anchor, positive, and negative tensors are expected to have "
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:269]: This should be renamed to `input`
```suggestion
            "The input, positive, and negative tensors are expected to have "
```

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 4]
```diff
@@ -230,3 +230,56 @@ def margin_ranking_loss(input, other, label, margin=0.0, reduction="mean", name=
...
+            f"the same number of dimensions, but got: anchor {a_dim}D, "
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:270]: This also should be set to `input`
```suggestion
            f"the same number of dimensions, but got: input {a_dim}D, "
```

File: ivy/functional/frontends/paddle/nn/functional/loss.py [hunk 5]
```diff
@@ -230,3 +230,56 @@ def margin_ranking_loss(input, other, label, margin=0.0, reduction="mean", name=
...
+    input,
+    positive,
+    negative,
+    margin=1.0,
+    p=2.0,
+    eps=1e-06,
+    swap=False,
+    reduction="mean",
+):
+    def pairwise_distance(x1, x2, *, p=2.0, eps=1e-06, keepdim=False):
```
Comment (hirwa-nshuti) [inline ivy/functional/frontends/paddle/nn/functional/loss.py:247]: This inner function better should be declared as a private function and should be better defined above this function and only be called inside this function

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_loss.py [hunk 6]
```diff
@@ -391,3 +391,55 @@ def test_paddle_margin_ranking_loss(
...
+    backend_fw,
+    frontend,
+    on_device,
+):
+    input_dtype, x = dtype_and_inputs
+    anchor_dtype, input = input_dtype[0], x[0]
+    positive_dtype, positive = input_dtype[1], x[1]
+    negative_dtype, negative = input_dtype[2], x[2]
+    helpers.test_frontend_function(
+        input_dtypes=[anchor_dtype, positive_dtype, negative_dtype],
```
Comment (hirwa-nshuti) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_loss.py:432]: It's fine to pass the `input_dtype` variable here not necessarily to unpack all those variables above

File: ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_loss.py [hunk 7]
```diff
@@ -391,3 +391,55 @@ def test_paddle_margin_ranking_loss(
...
+        backend_to_test=backend_fw,
+        frontend=frontend,
+        test_flags=test_flags,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        input=input,
```
Comment (hirwa-nshuti) [inline ivy_tests/test_ivy/test_frontends/test_paddle/test_nn/test_functional/test_loss.py:438]: input should be `x[0]` and other values should be assigned accordingly

**Review by hirwa-nshuti — CHANGES_REQUESTED** [review]
Hey @bundleabundance 

The tests are passing just suggested some refactoring then I'll get the PR merged😊

**Review by hirwa-nshuti — APPROVED** [review]
Looks great to me 😊

Thanks for contributing 🙂
